router: add default tls keypair to server

benburkert opened the pull request:
This is an initial attempt at implementing #267. I'm still getting up to speed on the code base so I'd love to get some feedback on this approach & any style issues.

The keypair passed into the new `HTTPListener` is used as the default `tls.Certificate` for SNI & non-SNI clients alike. This is handled by `tls.Config`'s `NameToCertificate` map, which is populated in the `tls.Server` call.

titanous commented on router/server.go:
```diff
@@ -67,9 +68,19 @@ func main() {
 	tcpRangeStart := flag.Int("tcp-range-start", 3000, "tcp port range start")
 	tcpRangeEnd := flag.Int("tcp-range-end", 3500, "tcp port range end")
+	certFile := flag.String("tlscert", "", "TLS (SSL) cert file in pem format")
+	keyFile := flag.String("tlskey", "", "TLS (SSL) key file in pem format")
```
This is good for anyone running the router standalone, but won't work inside of Flynn as there's no good way to inject config files into a container currently. We should support environment variables as well (which is what we're currently using for secrets in lieu of a more formal/secure system).

benburkert replied:
gotcha, any concerns with abusing `flag.String` to fallback to the environment variable?

```go
flag.String("tlscert", os.Getenv("TLSCERT"), "...")
```

benburkert replied:
also, `tlscert`/`tlskey` was the first name that popped into my head, lemme know if there's a better term for these.

titanous replied:
Will that fallback work? The environment variable will contain the data, but the flag is supposed to be a file path. The flag names are fine, but the description should note that they are for the default certificate presented by the HTTP listener (not the API, which is also HTTP).

benburkert replied:
ahh, the environment variables should the PEM encoded data, not the file path, gotcha.

titanous commented on router/http.go:
```diff
@@ -47,13 +47,14 @@ type HTTPListener struct {
 }
 
-func NewHTTPListener(addr, tlsAddr string, cookieKey *[32]byte, ds DataStore, discoverdc DiscoverdClient) *HTTPListener {
+func NewHTTPListener(addr, tlsAddr string, cookieKey *[32]byte, keypair tls.Certificate, ds DataStore, discoverdc DiscoverdClient) *HTTPListener {
```
This function is getting pretty overloaded. What do you think about moving to setting fields on the struct directly and then do a lazy init in `Start`?

benburkert replied:
sounds good. :+1:

titanous commented:
This looks good!

titanous commented:
I think a second component of #267 is allowing an arbitrary set of global wildcard and SAN certs to be used as a fallback for SNI connections, but that's much lower priority.

titanous commented:
FYI each commit needs a [DCO sign-off](https://flynn.io/docs/contributing#developer’s-certificate-of-origin) to make Travis happy.

titanous commented:
Oh, one other note that hasn't gotten added to the CONTRIBUTING document yet is that we are trying to add a bit of context to each commit message, as described here: https://groups.google.com/d/msg/golang-dev/6M4dmZWpFaI/SyU5Sl4zZLYJ

benburkert commented:
I added a test for a SAN domain (in the default keypair, not a route's keypair) and changed the listeners to use lazy init in the `Start` methods. The environment variable stuff still needs to be done, but this is ready for another :eyes:.

titanous commented on router/http.go:
```diff
@@ -88,6 +68,27 @@ func (s *HTTPListener) Close() error {
+	s.Watcher = s.wm
+
+	if s.ds == nil {
+		return errors.New("http listener missing data store")
```
Per convention, these errors should start with `router: `.

titanous commented on router/server.go:
```diff
@@ -106,9 +117,23 @@ func main() {
+			cookieKey: cookieKey,
+			keypair:   keypair,
+			ds:        NewEtcdDataStore(etcdc, path.Join(prefix, "http/")),
+			discoverd: d,
```
What do you think about 'exporting' these fields to distinguish them from the internal bookkeeping fields?

benburkert replied:
I agree that it's awkward, but i'd rather save this for a future refactor.

titanous replied:
K, that's fine.

benburkert commented:
This is ready for another :eyes:

titanous commented:
LGTM.

Please squash c13579302141aecbac7f6fa8f2f94236baeed7a0, and it looks like this needs a rebase to make travis happy.

titanous commented:
:sunrise:
